packages/backend/src/vm-manager.ts

```diff
@@ -0,0 +1,200 @@
+import fs from 'node:fs';
+import type { BootcBuildInfo } from '/@shared/src/models/bootc';
+
+// Singular pid file location (can only run 1 VM at a time)
+// eslint-disable-next-line sonarjs/publicly-writable-directories
+const pidFile = '/tmp/qemu-podman-desktop.pid';
+
+// MacOS related
+const macQemuArm64Binary = '/opt/homebrew/bin/qemu-system-aarch64';
+const macQemuArm64Edk2 = '/opt/homebrew/share/qemu/edk2-aarch64-code.fd';
+const macQemuX86Binary = '/opt/homebrew/bin/qemu-system-x86_64';
+
+// Default values for VM's
+const hostForwarding = 'hostfwd=tcp::2222-:22';
+const memorySize = '4G';
+const websocketPort = '45252';
+const rawImageLocation = 'image/disk.raw';
+
+export default class VMManager {
+  private build: BootcBuildInfo;
+
+  // Only values needed is the location of the VM file as well as the architecture of the image that
+  // will be used.
+  constructor(build?: BootcBuildInfo) {
+    this.build = build!;
+  }
+
+  // Launch the VM by generating the appropriate QEMU command and then launching it with process.exec
+  public async launchVM(): Promise<void> {
+    const diskImage = this.getDiskImagePath();
+
+    if (!fs.existsSync(diskImage)) {
+      throw new Error(`Raw disk image not found: ${diskImage}`);
+    }
+
+    try {
+      const command = this.generateLaunchCommand(diskImage);
+
+      if (command.length === 0) {
+        throw new Error(
+          'Unable to generate the launch command for the VM, ensure you are on the appropriate OS and architecture.',
+        );
+      }
+
+      await extensionApi.process.exec('sh', ['-c', `${command.join(' ')}`]);
+    } catch (e) {
+      this.handleError(e);
+    }
+  }
+
+  // We only support running one VM at at a time, so we kill the process by reading the pid from the universal pid file we use.
+  public async stopCurrentVM(): Promise<void> {
+    try {
+      await extensionApi.process.exec('sh', ['-c', `kill -9 \`cat ${pidFile}\``]);
+    } catch (e) {
+      // Ignore if it contains 'No such process' as that means the VM is already stopped / not running.
+      if (e instanceof Error && 'stderr' in e && typeof e.stderr === 'string' && e.stderr.includes('No such process')) {
+        return;
+      }
+      this.handleError(e);
+    }
+  }
+
+  // Prerequisite checks before launching the VM which includes checking if QEMU is installed as well as other OS specific checks.
+  public async checkVMLaunchPrereqs(): Promise<string | undefined> {
+    const diskImage = this.getDiskImagePath();
+    if (!fs.existsSync(diskImage)) {
+      return `Raw disk image not found at ${diskImage}. Please build a .raw disk image first.`;
+    }
+
+    if (!this.isArchitectureSupported()) {
+      return `Unsupported architecture: ${this.build.arch}`;
+    }
+
+    // Future support for Mac Intel, Linux ARM, Linux X86 and Windows ARM, Windows X86 to be added here.
+    if (isMac() && isArm()) {
+      return this.checkMacPrereqs();
+    } else {
```

Comment:
I think it doesn't scale well there, the idea was to have multiple instances

like you have an abstract base, then a MacArm, a MacIntel that has a preReqs method to fill
it's not the client that needs to do the if if at each call